### PR TITLE
Refactor list routes

### PR DESCRIPTION
**Pull request by phamann**
Apologies for the gigantic PR :(

This refactors the list routes to be more RESTful and most importantly consolidates the `/list/id` routes so we can serve public, private, RSS and JSON all of the same resource. A result of lots of discussion with the team this week.

### New routes:
- `GET /list` - gets create new list form
- `POST /list` - endpoint to post new list data
- `GET /list/:listId` - get individual list
- `GET /lists/:userId` - get all lists for user

### Includes:
- Route refactoring
- New auth middleware for public/private lists
- New create list redirection logic and messaging
- New view logic to show hide list actions when public/private
- Refactor tests

### Note:
I have had to leave the legacy routes to ensure we don't break functionality across the site whilst we update `myft-ui` to point to the new version of the routes.

### Todo:
- [ ] Add public/private toggle to lists view
- [x] Tests with @laurieboyes's "enhanced" work :panda_face:

### CODE REVIEW

File: server/controllers/lists/create.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+'use strict';
+
+const apiProxy = require('../../lib/api-proxy');
+const uuid = require('node-uuid').v4;
+const sanitizeHtml = require('../../lib/sanitize-html');
+
+module.exports = function (req, res, next) {
```
Comment (keirog) [inline server/controllers/lists/create.js]: Needs more hipster

File: server/controllers/lists/create.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+					actorId: req.body.uuid,
+					relationship: 'contained',
+					type: 'content',
+					method: 'post',
+					body: {
+						uuid: req.body.contentUuid
+					}
+				}, res.locals.flags)
+				.then(() => Promise.resolve('content'));
+				// .then(() => Promise.resolve(`You have added the article to the list. <a href="/myft/saved-articles/${res.locals.userId}/list/${req.body.uuid}">View list</a>`));
```
Comment (laurieboyes) [inline server/controllers/lists/create.js]: presuming these are dead comments?

File: server/controllers/lists/create.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+					actorName: 'list',
+					actorId: req.body.uuid,
+					relationship: 'contained',
+					type: 'content',
+					method: 'post',
+					body: {
+						uuid: req.body.contentUuid
+					}
+				}, res.locals.flags)
+				.then(() => Promise.resolve('content'));
```
Comment (laurieboyes) [inline server/controllers/lists/create.js]: I realise this was actually from my code, but do we definitely need to wrap the the returned value in another promise? I've got a feeling the return from a then() is already wrapped.

File: server/controllers/lists/create.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+				// .then(() => Promise.resolve(`You have added the article to the list. <a href="/myft/saved-articles/${res.locals.userId}/list/${req.body.uuid}">View list</a>`));
+
+			} else {
+				return Promise.resolve('list');
+				// return Promise.resolve(`You have created the list: <b>${req.body.name}</b>. <a href="/myft/lists">Manage lists</a>`);
+			}
+
+		})
+		.then(message => res.redirect(`/myft/list/${req.body.uuid}?message=${message}`))
+		.catch((err) => {
```
Comment (laurieboyes) [inline server/controllers/lists/create.js]: unnecessary parentheses

File: server/controllers/lists/form.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+const getLists = require('../../lib/lists');
+const sanitizeHtml = require('../../lib/sanitize-html');
+
+module.exports = function(req, res, next) {
+
+	return getLists(res.locals.userId)
+		.then(lists => {
+
+			let viewModel = {
```
Comment (laurieboyes) [inline server/controllers/lists/form.js]: could be const

**Comment by keirog** [comment]
LGTM

File: server/controllers/lists/read.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+module.exports = (req, res, next) => {
+	const render = renderStream(req, res, next);
+	const listId = req.params.listId;
+	const userId = res.locals.userId;
+
+	fetchListContent(listId)
+		.then(list => {
+
+			let listOwner = list.created[0].uuid;
+			let isOwner = userId === listOwner;
```
Comment (laurieboyes) [inline server/controllers/lists/read.js]: const

**Comment by laurieboyes** [comment]
:+1: :+1: :+1: apart from this PR https://github.com/Financial-Times/next-myft-page/pull/467 and pedantic comments

**Pull request merged**